File: 205. Isomorphic Strings/isIsomorphic.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
const isIsomorphic = (s, t) => {
  const mapST = {};
  const mapTS = {};

  for (let i = 0; i < s.length; i++) {
    let charS = s[i];
    let charT = t[i];

    if ((mapST[charS] !== undefined && mapST[charS] !== charT) || 
        (mapTS[charT] !== undefined && mapTS[charT] !== charS)) {
      return false;
    }

    mapST[charS] = charT;
    mapTS[charT] = charS;
  }

  return true;
};


const s = "egg";
const t = "add";
console.log('isIsomorphic: ', isIsomorphic(s, t))

File: 205. Isomorphic Strings/isIsomorphic.py
class Solution:
  def isIsomorphic(self, s: str, t: str) -> bool:
    mapST, mapTS = {}, {}
    
    for i in range(len(s)):
      charS, charT = s[i], t[i]
      
      if ((charS in mapST and mapST[charS] != charT) or
          (charT in mapTS and mapTS[charT] != charS)):
        return False
      
      mapST[charS] = charT
      mapTS[charT] = charS
    
    return True


s = "egg"
t = "add"
solution = Solution()
print('isIsomorphic: ', solution.isIsomorphic(s, t))
